### ISSUE TITLE
Fastai Failed to run CallbackHander.on_epoch_end logger

### ISSUE THREAD

**Issue by tdekelver** [comet-ml/issue-tracking#212]
Hi when I'm running some experiment on Google Colab server I get the following error message after training my model:
COMET INFO: Failed to run CallbackHander.on_epoch_end logger
Also the metrics defined does not seem to be added to the dashboard:

my code:
```%reload_ext autoreload
%autoreload 2
%matplotlib inline

from comet_ml import Experiment

from fastai.vision import *
from fastai import *

# Create an experiment
experiment = (...)
# get the data
path =...
arch = models.resnet50
tfms = get_transforms()
data = ...

# some custom loss function and metrics

learn = create_cnn(data, arch, metrics=[error_rate,accuracy,cohen_kappa(num_clas=len(data.classes)),MCC()]
                   , bn_final=True, loss_func = myFocalLoss(gamma = 2,cost = None))
#training
with experiment.train():
  learn.fit_one_cycle(1, slice(5e-2))
```

**Comment by dsblank** [comment]
Do you get the same results if you leave off the `%autoreload 2` ? That type of forced reloading might play havoc with our auto logging system.

**Comment by tdekelver** [comment]
I still have the same isssue if I remove that line...

**Comment by tdekelver** [comment]
I have done some tests and it does work with my custom loss function, but not when I add my custom metric. Then it fails.
This is the code I used to define my custom metric:
```
@dataclass
class my_metric(Callback):
    num_clas:int = 2

    def on_epoch_begin(self, **kwargs):
        self.sum_pred = torch.zeros(1, self.num_clas).float()
        self.sum_true = torch.zeros(1, self.num_clas).float()
        self.sum_false_pred = 0

    def on_batch_end(self, last_output, last_target, **kwargs):
        y_pred = last_output.argmax(dim=1).float()
        y_true = last_target.float()

        for i in range(self.num_clas):
            self.sum_pred[0, i] += (y_pred == i).float().sum()
            self.sum_true[0, i] += (y_true == i).float().sum()

        self.sum_false_pred += (y_pred != y_true).float().sum()

    def on_epoch_end(self, **kwargs):
        expected = torch.mm(self.sum_pred.view(self.num_clas, 1),
                            self.sum_true.view(1, self.num_clas)) / self.sum_pred.float().sum()
        mat_id = torch.matrix_power(torch.ones(self.num_clas, self.num_clas), 0) 
        w_mat = torch.ones(self.num_clas, self.num_clas) - mat_id
        k = self.sum_false_pred / (w_mat * expected).sum()

        self.metric = 1 - k
```
And I add it to my learner object by doing
`learn = create_cnn(data, arch, metrics=[error_rate,accuracy,my_metric(num_clas=len(data.classes))])`
Can you tell me why the CallbackHander.on_epoch_end logger fails to run when I add this custom metric and works when I remove it?

**Comment by tdekelver** [comment]
any update yet? would be really useful to make it work, as I really need to track the specific metric I defined and not the default ones!

**Comment by dsblank** [comment]
Sorry for the delay... this was a bit tricky to track down.

It turns out that we were relying on the convention that fastai metrics contain a `self.__name__` attribute. You can workaround the issue by adding this code:

```python
@dataclass
class my_metric(Callback):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.__name__ = "my_metric"
```

We will make our code more robust. Thank you for reporting and the followup. Please let us know if this doesn't solve the issue.

**Comment by tdekelver** [comment]
Thanks this solved the issue
PS I had to remove the `*args, **kwargs` from the `super().__init__` as such:

```python
@dataclass
class my_metric(Callback):
    def __init__(self, *args, **kwargs):
        super().__init__()
        self.__name__ = "my_metric"
```
